### PR TITLE
Add Unit Tests for Variant Class

### PR DESCRIPTION
**Pull request by sps1112**
- This PR is a part of issue #43440.
- These include the Assignment of one Variant class into another. 
- Also tests for the values.

Tests added for:-
- Conversion to Types:-
  - Bool
  - Int
  - Float
  - String
  - Vector2
  - Vector2i
  - Vector3
  - Vector3i
  - Color
- The conversion is from :- 
  - Bool
  - Int
  - Float
  - String
  - Vector2
  - Vector2i
  - Vector3
  - Vector3i
  - Color

### CODE REVIEW

File: tests/test_variant.h [hunk 1]
```diff
@@ -105,6 +105,600 @@ TEST_CASE("[Variant] Writer and parser float") {
...
 }
+
+TEST_CASE("[Variant] To Bool from Int,Float,String,Vec2,Vec2i,Vec3,Vec3i and Color") {
+	Variant int_v = 0;
+	Variant bool_v = true;
+	int_v = bool_v; // int_v is now a bool
+	CHECK(int_v == Variant(true));
+	bool_v = false;
+	int_v = bool_v;
+	CHECK(int_v == Variant(false));
```
Comment (Xrayez) [inline tests/test_variant.h:116]:
```suggestion
	CHECK(int_v == 0);
```

File: tests/test_variant.h [hunk 2]
```diff
@@ -105,6 +105,600 @@ TEST_CASE("[Variant] Writer and parser float") {
...
+
+	Variant float_v = 0.0f;
+	bool_v = true;
+	float_v = bool_v;
+	CHECK(float_v == Variant(true));
+	bool_v = false;
+	float_v = bool_v;
+	CHECK(float_v == Variant(false));
```
Comment (Xrayez) [inline tests/test_variant.h:124]:
```suggestion
	CHECK(float_v == 0.0);
```

File: tests/test_variant.h [hunk 3]
```diff
@@ -105,6 +105,600 @@ TEST_CASE("[Variant] Writer and parser float") {
...
+
+	Variant string_v = "";
+	bool_v = true;
+	string_v = bool_v;
+	CHECK(string_v == Variant(true));
```
Comment (Xrayez) [inline tests/test_variant.h:129]:
```suggestion
	CHECK(string_v == String("True"));
```

File: tests/test_variant.h [hunk 4]
```diff
@@ -105,6 +105,600 @@ TEST_CASE("[Variant] Writer and parser float") {
...
+	bool_v = false;
+	string_v = bool_v;
+	CHECK(string_v == Variant(false));
```
Comment (Xrayez) [inline tests/test_variant.h:132]:
```suggestion
	CHECK(string_v == String("False"));
```

And so on, the same applies to the rest of assertions.

If you'd like to see what values get converted to, you can use this snippet in GDScript:
```gdscript
extends Node2D

func _ready():
	var v = convert("", TYPE_BOOL) # Lookup TYPE_* in the documentation for other `Variant` types.
	print(typeof(v))
	print(v)
```

That said, you can also test `Variant::construct()` directly if you'd like to test conversion between various types this way.

File: tests/test_variant.h [hunk 5]
```diff
@@ -105,6 +105,600 @@ TEST_CASE("[Variant] Writer and parser float") {
...
+	CHECK(col_v == Variant(true));
+	bool_v = false;
+	col_v = bool_v;
+	CHECK(col_v == Variant(false));
+}
+
+TEST_CASE("[Variant] To Int from Bool,Float,String,Vec2,Vec2i,Vec3,Vec3i and Color") {
+	Variant bool_v = false;
+	Variant int_v = 2;
+	bool_v = int_v; // Now bool_v is int
```
Comment (Xrayez) [inline tests/test_variant.h:178]: You can check this explicitly with:
```suggestion
	bool_v = int_v;
	CHECK(bool_v.get_type() == Variant::TYPE_BOOL);
```

File: tests/test_variant.h [hunk 6]
```diff
@@ -105,6 +105,600 @@ TEST_CASE("[Variant] Writer and parser float") {
...
+	int_v = bool_v; // int_v is now a bool
+	CHECK(int_v == Variant(true));
+	bool_v = false;
+	int_v = bool_v;
+	CHECK(int_v == Variant(false));
+
+	Variant float_v = 0.0f;
+	bool_v = true;
+	float_v = bool_v;
+	CHECK(float_v == Variant(true));
```
Comment (Xrayez) [inline tests/test_variant.h:121]:
```suggestion
	CHECK(float_v == 1.0);
```

**Review by Xrayez — CHANGES_REQUESTED** [review]

**Comment by Xrayez** [comment]
I have deleted most of my reviews because I've misunderstood the purpose of those tests, sorry.

Those test cases check the `Variant` assignment, but not the actual conversion. If you'd like to see how values are converted, you can use this snippet in GDScript:
```gdscript
extends Node2D

func _ready():
	var v = convert("", TYPE_BOOL) # Lookup TYPE_* in the documentation for other `Variant` types.
	print(typeof(v))
	print(v)
```

That said, there's `Variant::construct()` (seen in `core/variant/variant_construct.cpp`) (equivalent to `convert()` in GDScript) to test conversion directly between various types.

So, I think current tests are fine, but I'd rename test cases to denote that those are `[Variant] Assignment` tests.

**Review by sps1112 — COMMENTED** [review]
I will be applying the requested changes shortly

File: tests/test_variant.h (hunk 5 above)
Comment (sps1112) [reply tests/test_variant.h:178]: I have added the explicit checking in place of the second Check Statement in each conversion. I have also kept the initial tests in the first Check Statement of each conversion just to check that the values are as expected.

**Review by Xrayez — APPROVED** [review]
Ok, tests pass so I'd merge. 🙂

You'll need to squash your commits into one by rebasing your PR, see https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase.

**Comment by sps1112** [comment]
Done.

**Comment by akien-mga** [comment]
Could you amend the commit message to be more explicit? "Update $filename" is not a good description for a commit, see https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-messages-with-readability-in-mind

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**